enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,374 @@
+| `description?: string` | The description of the plugin. |
+| `extensions: Extension[]` | List of extensions contributed by the plugin. |
+| `dependencies?: { [pluginName: string]: string }` | A plugin which depends on other plugins must explicitly define their dependencies. |
+
+##### Extension<P = JSON>
+
+| Property | Description |
+| ----- | -------- |
+| `id: string` | Unique identifier of this extension. |
+| `type: string` | The `type` identifies this extension of a particular extension type. |
+| `properties: P` | The extension properties JSON. |
+| `flags?: Partial<{ required: string[]; disallowed: string[]; }>` | Reference to the flags used to trigger the enablement of this extension. |
+
+An extension may contain code references, encoded as object literal `{
+$codeRef: string }`. The value of `$codeRef` is `moduleName.exportName` or
+`moduleName` (equivalent to `moduleName.default`). Console will resolve the
+code references when needed.
+
+### Lazy Loading
+
+Console uses
+[code splitting](https://webpack.js.org/guides/code-splitting/) to only load
+JavaScript code when needed. Likewise, dynamic plugin code will only load when
+needed. For instance, if the plugin contributes a list page, the plugin code
+for that list page only loads when the user visits the page. This improves
+performance and guards against bugs in plugins the user is not actively using.
+
+Console performs feature detection on initial load. This is used today to
+enable or disable parts of the UI based on features in the cluster and is
+typically driven by the resources found during API discovery. Dynamic plugins
+will also be able to declare console feature flags and will only load if the
+flag is enabled. For instance, the KubeVirt plugin will only load if the
+`VirtualMachine` resource is present.
+
+### Error Handling
+
+Console will guard against runtime errors in plugins. All plugin components
+will be surrounded by [React error boundaries](https://reactjs.org/docs/error-boundaries.html).
+This prevents an uncaught error from causing the application to white-screen
+and break. If a plugin service is unavailable, console will not load the plugin
+and log an error.
+
+Additionally, the console can contain a `?disable-plugins` query parameter.
+When this parameter is present, no dynamic plugins are loaded.
+
+### Risks and Mitigations
+
+**Risk**: Patternfly updates with breaking API changes can break plugins.
+
+**Mitigation**: Plugins can declare a semver range for compatibility. The core
+console team will coordinate breaking changes with the plugin teams so that a
+new operator version will be available with fixes when the corresponding
+OpenShift version GAs. We will not ship any breaking Patternfly changes in a
+z-stream.
+
+**Risk**: Plugins are not sandboxed and can make any API request on behalf of the user.
+
+**Mitigation**: Plugins are not enabled by default. A cluster admin must opt-in.
+
+## Design Details
+
+### Test Plan
```

Comment:
As we expose public API we need to have API test suites to ensure we maintain compatibility across releases.